### PR TITLE
check-meta.nix, lib/types.nix: type check changes, more documentation

### PR DESCRIPTION
**Pull request by piegamesde**
###### Description of changes

- Document that the `checkMeta` type check does not really work as intended
- Enable it by default nevertheless cause it's better than nothing, THIS IS A BREAKING CHANGE
- Small changes to lib/types.nix that are related

<del>- Enable metadata checking by default, see https://github.com/NixOS/nixpkgs/pull/25304#issuecomment-298385426
<del>- Check metadata before any other package issues, see https://github.com/NixOS/nixpkgs/issues/191124#issuecomment-1246523976
<del>- TODO fix soundness, see https://github.com/NixOS/nixpkgs/pull/6794#issuecomment-1003397767
  <del>- We really need to get that one fixed but I don't know too much about the internals, help appreciated here
<del>- Also, the error message is currently as helpful as "has a value of invalid type set; expected list of submodule", I'd really like to get them to the same quality as what we have in the NixOS module system.

CC @ncfavier @copumpkin @nbp 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ncfavier** [comment]
Second point LGTM.

I think the first point deserves its own PR as it's a potentially breaking change with a big performance impact (I haven't read the linked issue in its entirety, but the fact that there was this much discussion tells me that it's not a trivial change to make, even 5 years later, and probably needs more discussion).

For the third and fourth points, see my suggestion at https://github.com/NixOS/nixpkgs/pull/6794#discussion_r970730794.

**Comment by piegamesde** [comment]
As far as I can tell both the first and the third point are kind of breaking changes. As far as I can tell the default was left off in fear of breaking evaluation, but at some time we need to get on and break some eggs. It didn't automatically get better over five years, so we shouldn't expect it to happen in the future.

**Review by sternenseemann — CHANGES_REQUESTED** [review]
Overall, I don't like the module system type checking approach, I'd much rather just add new, stricter type variants to `lib.types` (e.g. `lib.types.strictListOf` or some better name) checks deeply and has a documented caveat about its performance.

File: pkgs/stdenv/generic/check-meta.nix
Comment (piegamesde) [reply pkgs/stdenv/generic/check-meta.nix:19]: Correct

File: pkgs/stdenv/generic/check-meta.nix
Comment (piegamesde) [reply pkgs/stdenv/generic/check-meta.nix:316]: Yes, this is from #177272, but I copied it over in order to test some more complex type checks. I'll remove it before merging.

**Comment by piegamesde** [comment]
Using the modules system is the least bad option I found: it takes care of default values, it can handle deep nesting, it gives us pretty decent error messages, it allows us to add example values and descriptions to meta attributes and I admit that I genuinely have no idea what half of the current values do.

I'm open to ideas on how to build type checks, but whatever we do they should work with composite types and I don't think this is currently possible without the module system or a lot of custom code. (The status quo simply iterates over all attributes, which totally ignores nested values)

File: pkgs/stdenv/generic/check-meta.nix
Comment (piegamesde) [reply pkgs/stdenv/generic/check-meta.nix]: To be honest I don't like a lot about that entire line. The usage of `${builtins.typeOf v}` leads to pretty bad error messages IMO ("Invalid type List expected listof str"). That's one of the reasons I'd like to avoid doing manual type checks like this and use the module system instead

File: pkgs/stdenv/generic/check-meta.nix
Comment (Profpatsch) [inline pkgs/stdenv/generic/check-meta.nix:302]: `.check` was wrong anyway, it doesn’t recurse into nested types.

File: pkgs/stdenv/generic/check-meta.nix
Comment (Profpatsch) [reply pkgs/stdenv/generic/check-meta.nix]: https://github.com/NixOS/nixpkgs/pull/180007

but @roberth didn’t like it

File: pkgs/stdenv/generic/check-meta.nix
Comment (roberth) [reply pkgs/stdenv/generic/check-meta.nix:302]: Only its name, `check`, is wrong. We need a function that evaluates as little as possible (WHNF) when checking definitions in the module system.

File: pkgs/stdenv/generic/check-meta.nix
Comment (roberth) [reply pkgs/stdenv/generic/check-meta.nix]: @Profpatsch My only dislike is that it adds too many parentheses. Those can be reduced a lot by implementing a few rules. I'll be happy to review it again if you or anyone improves it.

**Comment by tazjin** [comment]
> I'm open to ideas on how to build type checks, but whatever we do they should work with composite types and I don't think this is currently possible without the module system

Maybe you're talking about something else, but [yants](https://code.tvl.fyi/about/nix/yants) does not use the module system at all.

File: pkgs/stdenv/generic/check-meta.nix
Comment (SuperSandro2000) [inline pkgs/stdenv/generic/check-meta.nix:18]:
```suggestion
```

File: pkgs/stdenv/generic/check-meta.nix
Comment (SuperSandro2000) [inline pkgs/stdenv/generic/check-meta.nix:17]:
```suggestion
  # for why this defaulted to false. See also https://github.com/NixOS/nixpkgs/issues/191124 ~piegames
```

**Comment by SuperSandro2000** [comment]
@ofborg eval

File: pkgs/stdenv/generic/check-meta.nix
Comment (piegamesde) [reply pkgs/stdenv/generic/check-meta.nix]: Another reason against printing the value: It may blow up in the users face if they are accidentally putting some garbage in here, obscuring the actual error message.
https://github.com/NixOS/nixpkgs/pull/168295#issuecomment-1095518417

------

I like #180007, but it won't help against our usage of `${builtins.typeOf v}`

**Comment by piegamesde** [comment]
@tazjin yants looks really impressive and I'd love to see it in nixpkgs. The restriction for the current change is to use existing type checking infrastructure, and under that I don't see many good alternatives to a module system type check.

**Comment by Profpatsch** [comment]
I’d advise against setting `meta.check` to `true` by default, partially because ofborg runs the check. 

I’m afraid it will negatively influence eval times by ~a lot~, do we have benchmarks of disabled/enabled/enabled new impl?

**Comment by piegamesde** [comment]
I know of the performance cost issue, and even though I don't have exact numbers yet I assume that it will be noticeable. However, a type system that is not checked is worth nothing. So, either we do keep the type checks and also run them in our CI, or we go back to the usual untyped madness and remove `metaTypes` entirely. Sacrificing soundness for performance reasons like we do at the moment is not an option.

**Comment by 7c6f434c** [comment]
> However, a type system that is not checked is worth nothing.

That's false. It's useful documentation.

> Sacrificing soundness for performance reasons like we do at the moment is not an option.

If it were an actual policy, we would be quite limited in our compiler options for C++

Overloading OfBorg eval (which seems even harder to scale than builds) would be a much larger problem than what problems this change would solve.

**Comment by piegamesde** [comment]
I give up. I initially just stumbled upon this while implementing RFC 127 and tried to improve things a bit. I'd never planned to sink so much time into this, although the handful of closed previous attempts at fixing this should have warned me.

At the very least, I am going to document this madness so that the next person who runs into it knows what they are getting themselves into beforehand.

**Comment by Profpatsch** [comment]
For what it’s worth, I think that everything in here is an improvement over the status quo, except for the switch to check on every evaluation.

**Comment by piegamesde** [comment]
I stand by my decision to enable `checkMeta` by default, and there is always the option to opt-out in cases where it matters for performance.

File: pkgs/stdenv/generic/check-meta.nix
Comment (roberth) [reply pkgs/stdenv/generic/check-meta.nix:17]:
```suggestion
  shouldCheckMeta = config.checkMeta;
```

Could you add `checkMeta` to `pkgs/top-level/config.nix`? That way it's documented in the manual, including the default.

Comment (piegamesde) [reply pkgs/stdenv/generic/check-meta.nix]: Done.

**Comment by Profpatsch** [comment]
> I stand by my decision to enable `checkMeta` by default, and there is always the option to opt-out in cases where it matters for performance.

That doesn’t make any sense. The overhead comes from every package having to go through the module system type check, which is linear in the amount of `mkDerivation` calls you want to evaluate. So it’s either “enable for every evaluation” or “disable and only enable on CI”.

I’m strictly against flipping the switch before there are not some good benchmarks.

**Comment by piegamesde** [comment]
@Profpatsch Before stating opinions this strong, please take a look at the diff again. It has been updated yesterday and there is no module system involved anymore. (I think your statement is also wrong w.r.t. to the old changes because there was no module system involved either unless `checkMeta == true`, but there's no need discussing it further since that idea is off the table for now)

File: lib/types.nix
Comment (roberth) [inline lib/types.nix:122]:
```suggestion
      # Function applied to each definition that must return false when a definition
      # does not match the type. It should not check more than the root of the value,
      # because evaluating contained reduces laziness, leading to unnecessary
      # infinite recursions in the module system.
      # Further checks of contained values should be performed by throwing in
      # the merge function.
      # Strict and deep type checking can be performed by calling lib.deepSeq on
      # the merged value.
```

File: lib/types.nix
Comment (piegamesde) [reply lib/types.nix:419]: I initially had `let list = listOf anything;` locally, but it had a different description which I did not like. I decided to add it to `types.nix` to mirror the raw `attrs`, although I agree that we usually want to use neither of these.

But I can move it back to the let binding if you want to, no strong opinions on my side.

File: lib/types.nix
Comment (piegamesde) [reply lib/types.nix:116]: IMO it is, but I don't trust myself in renaming that function safely. But I'll remove the emoji while keeping the rest of the warning message.

File: nixos/doc/manual/release-notes/rl-2211.section.md
Comment (fricklerhandwerk) [inline nixos/doc/manual/release-notes/rl-2211.section.md:161]:
```suggestion
- The (previously undocumented) Nixpkgs configuration option `checkMeta` now defaults to `true`. This may cause evaulation
```

**Review by ncfavier — COMMENTED** [review]
So we're settling for a shallow check? That seems unsatisfactory and won't catch common mistakes like `maintainers = [ "J. Doe" ];`

Why is the deep check with the module system off the table? Is the performance impact big? Did anyone measure it? Would yants be better?

File: nixos/doc/manual/release-notes/rl-2211.section.md
Comment (ncfavier) [reply nixos/doc/manual/release-notes/rl-2211.section.md:161]: also "evaulation"

File: pkgs/stdenv/generic/check-meta.nix
Comment (piegamesde) [reply pkgs/stdenv/generic/check-meta.nix:261]: The only difference is that in the error message `listOf anything` would show instead of `list`, which I thought misleading. So I could equally do `list = (listOf anything) // { description = "list"; }` if you prefer to.

**Comment by piegamesde** [comment]
A deep check is not off the table and I'd love to see it eventually. But people had a lot of requirements for that –which is fine– and given that my main goal is to implement RFC 127 at the moment, I will not be the person implementing this. Therefore, I settled on properly documenting the the status quo which is the shallow check.

I added myself as CODEOWNER for that file and I will definitely review and help out whoever wants to tackle this in the future.

File: pkgs/stdenv/generic/check-meta.nix
Comment (ncfavier) [reply pkgs/stdenv/generic/check-meta.nix:261]: I think the error message should show the actual type, e.g. `list of string`, even if we don't check the "of string" part. In other words just keep the types as they were.

File: lib/types.nix
Comment (fricklerhandwerk) [inline lib/types.nix:122]:
```suggestion
    , # DO NOT USE WITHOUT KNOWING WHAT YOU ARE DOING!
```

File: lib/types.nix
Comment (fricklerhandwerk) [inline lib/types.nix:125]:
```suggestion
      # because checking nested values reduces laziness, leading to unnecessary
```

File: lib/types.nix
Comment (fricklerhandwerk) [inline lib/types.nix:127]:
```suggestion
      # Further checks of nested values should be performed by throwing in
```

... to keep consistency. I'm not religious about if you say nested or contained, just choose one - with with technical terms we should try to be unambiguous whether we mean the same thing by using the same words.

File: pkgs/stdenv/generic/check-meta.nix
Comment (fricklerhandwerk) [inline pkgs/stdenv/generic/check-meta.nix:248]:
```suggestion
  # A shallow type check. We are using NixOS'
```

**Review by fricklerhandwerk — APPROVED** [review]
LGTM in terms of docs. Only minor nit picks. Did not check the actual code.

**Comment by piegamesde** [comment]
@ofborg eval

**Comment by mweinelt** [comment]
Eval fix in #192891

**Comment by sternenseemann** [comment]
I wanna reiterate that this should not default to `true`. This causes quite annoying breakage downstream where completely harmless uses of the `meta` attribute in *dowstream* derivations are broken, e.g. adding new attributes for limited downstream usage. There is no reason why we need to pester downstream users with this when we check our packages in CI.

**Comment by sternenseemann** [comment]
Even worse, this apparently disallows a perfectly valid (in nixpkgs) extra attribute used for constructing channels in Hydra:

https://github.com/NixOS/nixpkgs/blob/350fd0044447ae8712392c6b212a18bdf2433e71/pkgs/build-support/release/default.nix#L116

This breaks every channel setup in Hydra instances.

**Comment by piegamesde** [comment]
@sternenseemann do you have some examples for custom downstream meta attributes?

**Comment by tazjin** [comment]
The TVL CI system (which is used by, well, TVL, and quite a few companies using Nix) uses `meta.ci` as the key under which CI for a target is controlled. As sterni points out, it's also used in Hydra instances.

Other Nix-based CI systems might also do something like this.

**Comment by piegamesde** [comment]
Okay, I see the following possibilities how to fix this:

- Tell people to disable `checkMeta` when evaluating packages with custom attributes
- Move all those custom attributes off `meta` into [`passthru`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome/extensions/buildGnomeExtension.nix#L53)
- Add some unchecked `metaPassthru` or `metaExtra` attribute
- Let `checkMeta` allow custom attributes in general
- Add `config.allowCustomMetaAttrs` or something which defaults to true
- Make `config.checkMeta` default to false again

**Comment by sternenseemann** [comment]
Note also that the current error message tells users absolutely nothing about how to get rid of it.

**Comment by tazjin** [comment]
For what it's worth, I think any workaround that requires changes to each package definition is bad and creates useless toil.

**Comment by aameen-tulip** [comment]
This change is brutal.

It overreaches into non-Nixpkgs just because Nixpkgs is used in an overlay or because `mkDerivation` is called.

FFS I have so much code that uses custom meta attributes, and it looks like I'm going to have our org pinned on the commit immediately before this for at least a few months while we clean up the red this put on the board...

**Comment by Profpatsch** [comment]
> I’d advise against setting `meta.check` to `true` by default, partially because ofborg runs the check.
> 
> I’m afraid it will negatively influence eval times by ~a lot, do we have benchmarks of disabled/enabled/enabled new impl?

Quoting myself here, I suggest we turn off default meta checks again. As far as I can see, no benchmarks regarding eval time have been done, either.

**Comment by piegamesde** [comment]
I'm [pretty sure OfBorg already runs the check](https://github.com/NixOS/ofborg/search?q=checkMeta) because when the default was false *somebody had to do the checks*. The performed checks are minimal compared to my initial proposal (iterate over one attrset with a shallow type check vs a full module system instantiation); benchmarks are not on my priority list right now.

Regarding the breakage this caused, this can easily be worked around by manually specifying `checkMeta = false;` until all outstanding issues have been resolved.

**Comment by Profpatsch** [comment]
> Regarding the breakage this caused, this can easily be worked around by manually specifying `checkMeta = false;` until all outstanding issues have been resolved.

Once again: No, that is not how upstream should act. Upstream shouldn’t break users and then tell them “just use a temporary solution until we get our shit together”.

As Linus states it: “We do not break Userspace. Period.”

**Comment by alyssais** [comment]
Nixpkgs has never agreed to guarantee stability of any API out-of-tree users might be using, and without any distinction between internal and public APIs, providing one would be infeasible.

I don't have an opinion on this change in particular, but Nixpkgs is not the kernel, and it's not even _like_ the kernel, due to the lack of a userspace/kernelspace distinction.

**Comment by sternenseemann** [comment]
Yes, thanks to changes like this, you have to treat nixpkgs as everything is kernel space, as it is often unreasonably hard to figure out changes without actively tracking development of nixpkgs itself. Case in point:

* nixpkgs has now started rejecting derivations defined downstream that it accepted before (even derivations produced by tools provided by nixpkgs itself, see #193310!).
* The feature that rejects the derivations is barely documented now and was completely undocumented before, so users don't know it.
* The error message does not tell you *what* part of nixpkgs causes this error and how to disable the check. Downstream users will not be aware that the check is optional, unless they dig through the backtrace or bisect nixpkgs.
* Even worse: There is no clear reason for forcing this check upon downstream users – having arbitrary extra stuff in `meta` is unproblematic.

**Comment by Profpatsch** [comment]
At least in this case it “just” causes evaluation failures, and not hard-to-debug runtime locale problems, but the annoyance exists nonetheless.

**Comment by piegamesde** [comment]
I'd like to point out that the reason why we are in this situation in the first place is that `meta` was an untyped mess without any checks for so long. Enabling security guards after the fact will never result in a smooth transition.

Going back and reverting this change is not an option. Instead, I'd like to have a constructive discussion about [ways to bring this forward](https://github.com/NixOS/nixpkgs/pull/191171#issuecomment-1260692375). I personally have no use cases that involve custom meta attributes, therefore I have no particular preferences for one above the other.

**Comment by Profpatsch** [comment]
> Going back and reverting this change is not an option.

Yes it is, it is precisely an option. One with a default value that we can flip.

**Comment by sternenseemann** [comment]
> I'd like to point out that the reason why we are in this situation in the first place is that `meta` was an untyped mess without any checks for so long. Enabling security guards after the fact will never result in a smooth transition.

This was only a problem for us, i.e. nixpkgs & nixos-search, no one else per se.

**Comment by Profpatsch** [comment]
> > Going back and reverting this change is not an option.
> 
> Yes it is, it is precisely an option. One with a default value that we can flip.

I’ve done it in https://github.com/NixOS/nixpkgs/pull/194058 which I strongly suggest we apply until further testing is in place.

**Comment by 7c6f434c** [comment]
> Enabling security guards after the fact will never result in a smooth transition.

Can you please explain how it is a _security_ guard? What issue is prevented by forbidding extra attributes in `meta`?

**Comment by piegamesde** [comment]
I'm unsubscribing from the discussion now. I spent a considerable amount of time trying to improve things, only to be met by – mostly unconstructive, IMO – criticism, and then ended up spending a multiple of that time in unproductive discussions. I am saddened by the fact that people prefer to cling to the status quo instead of envisioning a solution that allows `checkMeta` to be enabled by default *and* custom meta attributes outside of nixpkgs.

This is not the first time this has happened to me in nixpkgs, and the next time I start some project in the attempt to improve things I will think about it twice beforehand.

**Comment by Profpatsch** [comment]
I can tell you exactly how to make this change without breaking downstream, but I don’t think I was ever asked. Rather, my objections were ignored as irrelevant and then people started complaining that their setups broke.

Some approaches:

* Don’t flip the flag outright, set it to `"warn"` and display a warning when the typecheck fails, with a link to some nice instructions how to fix this.
* Change the check to allow extra fields in `meta`; this could potentially clash with future field additions of course, but it would not break any downstream right now and could be refined at a later date.

**Comment by mweinelt** [comment]
The proposal made in https://github.com/NixOS/nixpkgs/pull/191171#issuecomment-1260692375 were ignored all the same.

**Comment by tazjin** [comment]
It seems like

> Let checkMeta allow custom attributes in general

Is the same as 

> Change the check to allow extra fields in meta

Which seems pretty sane to me, so maybe we should do that?

**Comment by mweinelt** [comment]
As mentioned by @alyssais in https://github.com/NixOS/nixpkgs/pull/194058#issuecomment-1264645805 we would like to have stronger guarantees for nixpkgs.

> > Wouldn't it be sufficient to allow extra arguments in meta and enforce checks only on well-defined keys used throughout nixpkgs?
> 
> Upon reflection I don't think this is a good idea. The main use I've had of the meta checks in the past has actually been catching misspelled keys. So I think we really need a config option with an allow list of unrecognized meta checks.

**Pull request merged**
